src/mqt/qecc/circuit_synthesis/synthesis_utils.py

```diff
@@ -90,73 +94,452 @@ def iterative_search_with_timeout(
+
+    """
+    # NOTE: This implementation assumes single error events only. For higher-order errors, adjustments are required.
+    control, target = cnot_gate
+
+    control_row = propagation_matrix[control]
+    target_row = propagation_matrix[target]
+
+    if x_error:
+        propagation_matrix[control] = np.bitwise_xor(control_row, target_row)
+        return propagation_matrix[control], propagation_matrix
 
-    The algorithm utilizes a greedy heuristic to select the columns to eliminate in order to minimize the number of eliminations required.
+    propagation_matrix[target] = np.bitwise_xor(control_row, target_row)
+    return propagation_matrix[target], propagation_matrix
 
-    The matrix is reduced until there are exactly rnk(matrix) columns with non-zero entries.
+
+def check_mutually_disjointness_spcs(
+    c_spcs: list[StatePrepCircuit], p_spcs: list[StatePrepCircuit], x_error: bool = True
+) -> list[StatePrepCircuit]:
+    """Check if potential SPCs have mutually disjoint fault set with all current SPCs.
+
+    Take list of SPCs with already mutually disjoint fault sets and another list of potential candidates for new SPCs that
+    have a mutually disjoint fault set with all current SPCs.
 
     Args:
-        matrix: The matrix to perform Gaussian elimination on.
-        parallel_elimination: Whether to prioritize elimination steps that act on disjoint columns.
+        c_spcs: current SPCs that are already mutually disjoint in terms of fault sets
+        p_spcs: potential SPCs that shall be tested for mutually disjointness against the current set
+        x_error: flag that decides if check happens for x errors or z errors
+    """
+    i = 0
+    while i < len(p_spcs):
+        pspc = p_spcs[i]
+        # NOTE: check if it is not better to use the big fault set of the current SPC here as those sets have already been
+        # calculated and do not need to be calculated again
+        # FIX: big fs calculation should be only done for the current elements if there are more candidates than current elements
+        px_fs, pz_fs = get_fs_based_on_d(pspc)
+
+        # Check against *all* existing and newly added elements in c_spcs
+        for _spc in c_spcs:
+            # NOTE: for distance 5 and 7 we only compare against the 1 error fault set.
+            # This might need adjustments for different distances.
+            x_fs = _spc.compute_fault_set()
+            z_fs = _spc.compute_fault_set(x_errors=False)
+            if x_error:
+                if _spc.code.check_fs_overlap(x_fs, px_fs):
+                    break
+            elif _spc.code.check_fs_overlap(z_fs, pz_fs, x_error=False):
+                break  # Stop checking if there's an overlap
+        else:  # No overlap found, so add pspc
+            c_spcs.append(pspc)
+
+        i += 1  # Move to next element in p_spcs
+    return c_spcs
+
+
+def get_fs_based_on_d(spc: StatePrepCircuit) -> tuple[npt.NDArray[np.int8], npt.NDArray[np.int8]]:
```

Comment:
I am realizing more and more that dealing with fault sets should be better supported since it is an essential primitive in a lot of our synthesis tasks. Do you think it makes sense to add more support for fault sets? Nothing for this PR but maybe we can open an issue and deal with that somewhat soon.

Regarding this method specifically, I would argue that the calling context knows the distance. The logic between the d=3,5,7 cases are different enough that you could have a dedicated method for each and avoid the branching here.